[WPE] WPE Platform: Fix invisible cursor on wayland platform

Scony opened the pull request:
#### 474b00fa093a02cefdabc6d55750d46398ed1390
<pre>
[WPE] WPE Platform: Fix invisible cursor on wayland platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=272386">https://bugs.webkit.org/show_bug.cgi?id=272386</a>

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

* Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
(WPE::cursorsPath):
(WPE::CursorTheme::create):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
(WPE::WaylandCursor::WaylandCursor):
(WPE::WaylandCursor::setFromName):
(WPE::WaylandCursor::setFromBuffer):
(WPE::WaylandCursor::update const):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.h:

Canonical link: <a href="https://commits.webkit.org/277726@main">https://commits.webkit.org/277726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25c021108e3d23bc277fe6b4086028947487b487

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48407 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51365 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51095 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44472 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25141 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48989 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25303 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20691 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22787 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6463 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44727 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53000 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23454 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24719 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/41988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45798 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25524 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6888 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24442 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
```diff
@@ -31,10 +31,19 @@
+{
+    auto theme = WaylandCursorTheme::create(wpe_display_wayland_get_wl_shm(display));
+    if (theme)
+        return makeUnique<WaylandCursor>(display, WTFMove(theme));
```
Typically we write this kind of thing with the assignment inside the `if`:

```suggestion
    if (auto theme = WaylandCursorTheme::create(wpe_display_wayland_get_wl_shm(display)))
        return makeUnique<WaylandCursor>(display, WTFMove(theme));
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
```diff
@@ -31,10 +31,19 @@
+        return makeUnique<WaylandCursor>(display, WTFMove(theme));
+    g_warning("Could not load cursor theme, disabling custom cursors support");
+    return nullptr;
+}
```
I'm, not sure we want this, we are already handling the case of the theme being nullptr in WaylandCursor, with this patch we don't allow to set custom cursors when the theme fails to load, for example. I think we should avoid setting the cursor when it was set by a name and the theme failed to load. It would be interesting to know why the theme failed to load, and maybe we could even add a few embedded cursors and have a simple fallback theme to ensure we always have cursors

aperezdc reviewed (approved):
Another alternative would be to try using the [cursor-shape-v1](https://wayland.app/protocols/cursor-shape-v1) protocol, and fall back to have the client load cursor images (which is the current implementation). Anyway, that would rather be a follow-up.

Patch LGTM with one minor nit, please take a look before landing it.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
```diff
@@ -36,6 +36,8 @@ WaylandCursor::WaylandCursor(WPEDisplayWayland* display)
     , m_theme(WaylandCursorTheme::create(wpe_display_wayland_get_wl_shm(m_display)))
 {
+    if (!m_theme)
+        g_warning("Could not load cursor theme, disabling custom cursors support");
```
Custom cursors should work, it's named cursors what won't work.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
```diff
@@ -92,6 +94,8 @@ void WaylandCursor::setFromBuffer(struct wl_buffer* buffer, uint32_t width, uint
 void WaylandCursor::update() const
 {
+    if (!m_theme)
+        return;
```
This still breaks custom cursors. I think we need another member variable to set whether the cursor changed to return early here.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
```diff
@@ -69,6 +71,7 @@ void WaylandCursor::setFromName(const char* name, double scale)
 
     m_hotspot.x = cursor[0].hotspotX;
     m_hotspot.y = cursor[0].hotspotY;
+    m_cursorChanged = true;
```
We should set it to true when it's none too

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
```diff
@@ -44,10 +45,14 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
     };
 
-    GUniquePtr<char> baseCursorsPath(g_build_filename(basePath, "cursors", nullptr));
+    String canonicalPathOfIndex = WTF::FileSystemImpl::realPath(String::fromUTF8(GUniquePtr<char>(g_build_filename(basePath, "index.theme", nullptr)).get()));
```
Maybe we could do something like:

```
auto indexFilename = FileSystem::pathByAppendingComponent(String::fromUTF8(basePath), "index.theme"_s);
String canonicalPathOfIndex = FileSystem::realPath(indexFilename);
```

Scony committed:
[WPE] WPE Platform: Fix invisible cursor on wayland platform
https://bugs.webkit.org/show_bug.cgi?id=272386

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

* Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
(WPE::cursorsPath):
(WPE::CursorTheme::create):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.cpp:
(WPE::WaylandCursor::WaylandCursor):
(WPE::WaylandCursor::setFromName):
(WPE::WaylandCursor::setFromBuffer):
(WPE::WaylandCursor::update const):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandCursor.h:

Canonical link: https://commits.webkit.org/277726@main
